File: native/agent/src/process_snapshot.cpp
#include "process_snapshot.h"

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cctype>
#include <fstream>
#include <limits>
#include <sstream>
#include <system_error>
#include <unordered_map>
#include <unordered_set>

#include <unistd.h>

namespace fs = std::filesystem;

namespace mini_drop_native {
namespace {

std::atomic<std::uint64_t> g_observation_generation{0};

std::string bounded_text(std::string value, std::size_t limit) {
  value.erase(std::remove_if(value.begin(), value.end(), [](unsigned char ch) {
    return ch == 0 || ch == '\n' || ch == '\r' || ch < 0x20 || ch == 0x7f;
  }), value.end());
  if (value.size() > limit) value.resize(limit);
  return value;
}

bool numeric_name(const std::string& value) {
  return !value.empty() && std::all_of(value.begin(), value.end(), [](unsigned char ch) {
    return std::isdigit(ch) != 0;
  });
}

bool parse_u32(const std::string& value, std::uint32_t& result) {
  try {
    std::size_t parsed = 0;
    const std::uint64_t raw = std::stoull(value, &parsed);
    if (parsed != value.size() || raw == 0 ||
        raw > std::numeric_limits<std::uint32_t>::max()) {
      return false;
    }
    result = static_cast<std::uint32_t>(raw);
    return true;
  } catch (...) {
    return false;
  }
}

std::string read_first_line(const fs::path& path, std::size_t limit) {
  std::ifstream input(path);
  std::string line;
  if (!std::getline(input, line)) return {};
  return bounded_text(std::move(line), limit);
}

std::uint64_t inode_from_link(const fs::path& path) {
  std::error_code error;
  const std::string link = fs::read_symlink(path, error).string();
  if (error) return 0;
  const auto left = link.find('[');
  const auto right = link.find(']', left == std::string::npos ? 0 : left + 1);
  if (left == std::string::npos || right == std::string::npos) return 0;
  try {
    std::size_t parsed = 0;
    const std::string value = link.substr(left + 1, right - left - 1);
    const std::uint64_t inode = std::stoull(value, &parsed);
    return parsed == value.size() ? inode : 0;
  } catch (...) {
    return 0;
  }
}

bool parse_stat(
    const fs::path& path,
    std::uint32_t& parent_pid,
    std::uint64_t& start_ticks) {
  std::ifstream input(path);
  std::string line;
  if (!std::getline(input, line)) return false;
  const auto close = line.rfind(')');
  if (close == std::string::npos || close + 2 >= line.size()) return false;
  std::istringstream fields(line.substr(close + 2));
  std::string value;
  // The substring begins at field 3. PPID is field 4; start time is field 22.
  for (int field = 3; field <= 22; ++field) {
    if (!(fields >> value)) return false;
    if (field == 4) {
      try {
        std::size_t parsed = 0;
        const std::uint64_t raw = std::stoull(value, &parsed);
        if (parsed != value.size() ||
            raw > std::numeric_limits<std::uint32_t>::max()) {
          return false;
        }
        parent_pid = static_cast<std::uint32_t>(raw);
      } catch (...) {
        return false;
      }
    } else if (field == 22) {
      try {
        std::size_t parsed = 0;
        start_ticks = std::stoull(value, &parsed);
        if (parsed != value.size()) return false;
      } catch (...) {
        return false;
      }
    }
  }
  return start_ticks != 0;
}

std::uint32_t namespace_pid(const fs::path& status_path) {
  std::ifstream input(status_path);
  std::string line;
  while (std::getline(input, line)) {
    if (line.rfind("NSpid:", 0) != 0) continue;
    std::istringstream values(line.substr(6));
    std::uint64_t item = 0;
    std::uint64_t last = 0;
    bool found = false;
    while (values >> item) {
      last = item;
      found = true;
    }
    if (found && last > 0 && last <= std::numeric_limits<std::uint32_t>::max()) {
      return static_cast<std::uint32_t>(last);
    }
    return 0;
  }
  return 0;
}

std::string executable_identity(const fs::path& exe_link) {
  std::error_code error;
  fs::path target = fs::read_symlink(exe_link, error);
  if (error) return {};
  std::string identity = target.filename().string();
  if (identity.empty()) identity = target.string();
  return bounded_text(std::move(identity), kMaxProcessIdentityLength);
}

std::string cgroup_hint(const fs::path& path) {
  std::ifstream input(path);
  std::string line;
  std::string fallback;
  while (std::getline(input, line)) {
    const auto first_colon = line.find(':');
    if (first_colon == std::string::npos) continue;
    const auto second_colon = line.find(':', first_colon + 1);
    if (second_colon == std::string::npos) continue;
    std::string group = line.substr(second_colon + 1);
    if (group.empty() || group == "/") continue;
    if (fallback.empty()) fallback = group;
    if (group.find(".service") != std::string::npos ||
        group.find("docker") != std::string::npos ||
        group.find("kubepods") != std::string::npos ||
        group.find("containerd") != std::string::npos) {
      return bounded_text(std::move(group), kMaxProcessHintLength);
    }
  }
  return bounded_text(std::move(fallback), kMaxProcessHintLength);
}

std::string service_hint_from_cgroup(const std::string& cgroup) {
  std::size_t end = 0;
  while (end < cgroup.size()) {
    const std::size_t start = cgroup.find_first_not_of('/', end);
    if (start == std::string::npos) break;
    end = cgroup.find('/', start);
    const std::string part = cgroup.substr(
        start, end == std::string::npos ? std::string::npos : end - start);
    if (part.size() > 8 && part.compare(part.size() - 8, 8, ".service") == 0) {
      return bounded_text(part, kMaxProcessHintLength);
    }
    if (end == std::string::npos) break;
  }
  return {};
}

std::string instance_hint_from_cgroup(const std::string& cgroup) {
  std::string fallback;
  std::size_t end = 0;
  while (end < cgroup.size()) {
    const std::size_t start = cgroup.find_first_not_of('/', end);
    if (start == std::string::npos) break;
    end = cgroup.find('/', start);
    const std::string part = cgroup.substr(
        start, end == std::string::npos ? std::string::npos : end - start);
    if (!part.empty()) fallback = part;
    if (part.find("docker-") == 0 || part.find("cri-containerd-") == 0 ||
        part.find("crio-") == 0) {
      return bounded_text(part, kMaxProcessHintLength);
    }
    if (end == std::string::npos) break;
  }
  return bounded_text(std::move(fallback), kMaxProcessHintLength);
}

std::vector<std::string> bounded_capabilities(
    const std::vector<std::string>& capabilities) {
  std::vector<std::string> result;
  result.reserve(std::min(capabilities.size(), kMaxCollectorCapabilities));
  for (const auto& capability : capabilities) {
    if (result.size() == kMaxCollectorCapabilities) break;
    const std::string safe = bounded_text(capability, kMaxCollectorCapabilityLength);
    if (!safe.empty()) result.push_back(safe);
  }
  return result;
}

bool is_agent_process(
    std::uint32_t pid,
    std::uint32_t self_pid,
    const std::unordered_map<std::uint32_t, std::uint32_t>& parents) {
  std::unordered_set<std::uint32_t> visited;
  while (pid != 0 && visited.insert(pid).second) {
    if (pid == self_pid) return true;
    const auto parent = parents.find(pid);
    if (parent == parents.end()) return false;
    pid = parent->second;
  }
  return false;
}

}  // namespace

ProcessSnapshot collect_process_snapshot(
    const std::vector<std::string>& available_collectors,
    const fs::path& proc_root,
    std::size_t max_candidates,
    std::uint32_t self_pid) {
  ProcessSnapshot snapshot;
  snapshot.generation = ++g_observation_generation;
  snapshot.observed_at_unix_ms = static_cast<std::uint64_t>(
      std::chrono::duration_cast<std::chrono::milliseconds>(
          std::chrono::system_clock::now().time_since_epoch()).count());
  snapshot.boot_id = read_first_line(
      proc_root / "sys/kernel/random/boot_id", kMaxProcessIdentityLength);
  if (snapshot.boot_id.empty()) {
    snapshot.error = "cannot read boot_id";
    return snapshot;
  }
  if (self_pid == 0) self_pid = static_cast<std::uint32_t>(::getpid());
  max_candidates = std::min(max_candidates, kMaxProcessCandidates);
  const auto capabilities = bounded_capabilities(available_collectors);

  std::error_code error;
  fs::directory_iterator entries(proc_root, error);
  if (error) {
    snapshot.error = bounded_text(
        "cannot scan proc root: " + error.message(), kMaxProcessSnapshotErrorLength);
    return snapshot;
  }

  std::vector<fs::path> process_roots;
  for (fs::directory_iterator end; entries != end;) {
    if (error) {
      snapshot.error = "proc directory iteration failed";
      return snapshot;
    }
    if (numeric_name(entries->path().filename().string())) {
      process_roots.push_back(entries->path());
    }
    entries.increment(error);
  }
  if (error) {
    snapshot.error = "proc directory iteration failed";
    return snapshot;
  }
  std::sort(process_roots.begin(), process_roots.end(), [](const fs::path& left,
                                                            const fs::path& right) {
    std::uint32_t left_pid = 0;
    std::uint32_t right_pid = 0;
    parse_u32(left.filename().string(), left_pid);
    parse_u32(right.filename().string(), right_pid);
    return left_pid < right_pid;
  });

  struct ObservedProcess {
    fs::path root;
    std::uint32_t pid = 0;
    std::uint32_t parent_pid = 0;
    std::uint64_t start_ticks = 0;
  };
  std::vector<ObservedProcess> observed;
  std::unordered_map<std::uint32_t, std::uint32_t> parents;
  observed.reserve(process_roots.size());
  for (const auto& process_root : process_roots) {
    ObservedProcess process;
    process.root = process_root;
    if (!parse_u32(process_root.filename().string(), process.pid) ||
        !parse_stat(process_root / "stat", process.parent_pid, process.start_ticks)) {
      // Processes can exit during a scan. Omit identities we cannot attest.
      continue;
    }
    parents[process.pid] = process.parent_pid;
    observed.push_back(std::move(process));
  }

  for (const auto& process : observed) {
    if (is_agent_process(process.pid, self_pid, parents)) continue;

    ProcessCandidate candidate;
    candidate.pid = process.pid;
    candidate.process_start_ticks = process.start_ticks;
    candidate.pid_namespace_inode = inode_from_link(process.root / "ns/pid");
    candidate.namespace_pid = namespace_pid(process.root / "status");
    // Never attest a candidate without all immutable identity components.
    if (candidate.process_start_ticks == 0 || candidate.pid_namespace_inode == 0 ||
        candidate.namespace_pid == 0) {
      continue;
    }
    if (snapshot.candidates.size() == max_candidates) {
      snapshot.truncated = true;
      break;
    }

    candidate.comm = read_first_line(
        process.root / "comm", kMaxProcessIdentityLength);
    candidate.executable_identity = executable_identity(process.root / "exe");
    candidate.cgroup = cgroup_hint(process.root / "cgroup");
    candidate.service_hint = service_hint_from_cgroup(candidate.cgroup);
    candidate.instance_hint = instance_hint_from_cgroup(candidate.cgroup);
    candidate.collector_capabilities = capabilities;
    snapshot.candidates.push_back(std::move(candidate));
  }

  snapshot.complete = !snapshot.truncated;
  return snapshot;
}

}  // namespace mini_drop_native

File: agent/mini_drop_agent/process_snapshot.py
"""Authoritative host-process snapshots attached to Agent heartbeats.

The immutable identity fields collected here mirror the native Agent.  They
let the server bind a diagnosis to the process that was actually observed,
instead of trusting a reusable PID supplied by a browser.
"""

from __future__ import annotations

import os
import re
import socket
import time
from pathlib import Path
from typing import Iterable


MAX_CANDIDATES = 256
_HEX_CONTAINER_HOSTNAME = re.compile(r"[0-9a-f]{12,64}", re.IGNORECASE)


def _read_text(path: Path, limit: int = 256) -> str:
    try:
        value = path.read_text(encoding="utf-8", errors="replace").splitlines()[0]
    except (OSError, IndexError):
        return ""
    return "".join(ch for ch in value if ch >= " " and ch != "\x7f")[:limit]


def _parse_stat(path: Path) -> tuple[int, int] | None:
    try:
        value = path.read_text(encoding="utf-8", errors="replace")
        fields = value[value.rfind(")") + 2 :].split()
        parent_pid = int(fields[1])  # fields begin at Linux /proc stat field 3
        start_ticks = int(fields[19])
    except (OSError, ValueError, IndexError):
        return None
    return (parent_pid, start_ticks) if start_ticks > 0 else None


def _namespace_pid(path: Path) -> int:
    try:
        lines = path.read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError:
        return 0
    for line in lines:
        if line.startswith("NSpid:"):
            try:
                return int(line.split()[-1])
            except (ValueError, IndexError):
                return 0
    return 0


def _namespace_inode(path: Path) -> int:
    try:
        target = os.readlink(path)
        return int(target[target.index("[") + 1 : target.index("]")])
    except (OSError, ValueError):
        return 0


def _cgroup(path: Path) -> str:
    fallback = ""
    try:
        lines = path.read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError:
        return ""
    for line in lines:
        fields = line.split(":", 2)
        if len(fields) != 3 or fields[2] in {"", "/"}:
            continue
        fallback = fallback or fields[2]
        if any(token in fields[2] for token in (".service", "docker", "kubepods", "containerd")):
            return fields[2][:256]
    return fallback[:256]


def _hints(root: Path, cgroup: str, host_hostname: str) -> tuple[str, str]:
    parts = [part for part in cgroup.split("/") if part]
    service = next((part for part in parts if part.endswith(".service")), "")
    instance = parts[-1] if parts else ""
    container_hostname = _read_text(root / "root/etc/hostname")
    if (
        container_hostname
        and container_hostname != host_hostname
        and not _HEX_CONTAINER_HOSTNAME.fullmatch(container_hostname)
    ):
        service = container_hostname
        instance = container_hostname
    return service[:256], instance[:256]


def _is_agent_descendant(pid: int, self_pid: int, parents: dict[int, int]) -> bool:
    visited: set[int] = set()
    while pid and pid not in visited:
        if pid == self_pid:
            return True
        visited.add(pid)
        pid = parents.get(pid, 0)
    return False


def fill_process_candidate_snapshot(
    wire_snapshot,
    capabilities: Iterable[str],
    *,
    proc_root: str | Path = "/proc",
    self_pid: int | None = None,
) -> None:
    """Fill a protobuf ProcessCandidateSnapshot from one bounded /proc scan."""

    root = Path(proc_root)
    wire_snapshot.generation = time.monotonic_ns()
    wire_snapshot.observed_at_unix_ms = int(time.time() * 1000)
    wire_snapshot.boot_id = _read_text(root / "sys/kernel/random/boot_id")
    if not wire_snapshot.boot_id:
        wire_snapshot.error = "cannot read boot_id"
        return

    try:
        process_roots = sorted(
            (item for item in root.iterdir() if item.name.isdigit()),
            key=lambda item: int(item.name),
        )
    except OSError as exc:
        wire_snapshot.error = f"cannot scan proc root: {exc}"[:256]
        return

    observed: list[tuple[Path, int, int]] = []
    parents: dict[int, int] = {}
    for process_root in process_roots:
        parsed = _parse_stat(process_root / "stat")
        if parsed is None:
            continue
        pid = int(process_root.name)
        parent_pid, start_ticks = parsed
        parents[pid] = parent_pid
        observed.append((process_root, pid, start_ticks))

    safe_capabilities = [str(item)[:64] for item in capabilities][:16]
    own_pid = self_pid or os.getpid()
    host_hostname = _read_text(root / "1/root/etc/hostname") or socket.gethostname()
    for process_root, pid, start_ticks in observed:
        if _is_agent_descendant(pid, own_pid, parents):
            continue
        namespace_inode = _namespace_inode(process_root / "ns/pid")
        namespace_pid = _namespace_pid(process_root / "status")
        if not namespace_inode or not namespace_pid:
            continue
        try:
            executable_identity = Path(os.readlink(process_root / "exe")).name[:256]
        except OSError:
            # Kernel threads and processes that disappear mid-scan cannot be
            # bound immutably; omitting them keeps the snapshot authoritative.
            continue
        if not executable_identity:
            continue
        if len(wire_snapshot.candidates) >= MAX_CANDIDATES:
            wire_snapshot.truncated = True
            return
        candidate = wire_snapshot.candidates.add()
        candidate.pid = pid
        candidate.process_start_ticks = start_ticks
        candidate.pid_namespace_inode = namespace_inode
        candidate.namespace_pid = namespace_pid
        candidate.comm = _read_text(process_root / "comm")
        candidate.executable_identity = executable_identity
        candidate.cgroup = _cgroup(process_root / "cgroup")
        candidate.service_hint, candidate.instance_hint = _hints(
            process_root, candidate.cgroup, host_hostname
        )
        candidate.collector_capabilities.extend(safe_capabilities)

    wire_snapshot.complete = True
